### DISCUSSION TITLE
config script has empty target

### DISCUSSION THREAD

**Discussion by 8W9aG** [openssl/openssl#22826] [Q&A] [answered]
I am currently trying to get the openssl Configure script to render the header templates from bazel, I have the current genrule:

```
OPENSSL_CONFIG_CMD = """
OPENSSL_ROOT=$$(dirname $(location config))
pushd $$OPENSSL_ROOT
    ./config
popd
cp $$OPENSSL_ROOT/include/openssl/configuration.h $(location include/openssl/configuration.h)
"""

genrule(
    name = "configure",
    outs = GENERATED_OPENSSL_HEADERS,
    cmd = OPENSSL_CONFIG_CMD,
    tools = [
        "config",
        "configdata.pm.in",
        "Configure",
        "VERSION.dat",
        "util/perl/OpenSSL/config.pm",
        "util/perl/OpenSSL/fallback.pm",
        "util/perl/OpenSSL/Glob.pm",
        "util/perl/OpenSSL/Template.pm",
        "external/perl/Downloaded.txt",
        "external/perl/MODULES.txt",
        "external/perl/Text-Template-1.56/Changes",
        "external/perl/Text-Template-1.56/INSTALL",
        "external/perl/Text-Template-1.56/LICENSE",
        "external/perl/Text-Template-1.56/Makefile.PL",
        "external/perl/Text-Template-1.56/MANIFEST",
        "external/perl/Text-Template-1.56/META.json",
        "external/perl/Text-Template-1.56/META.yml",
        "external/perl/Text-Template-1.56/README",
        "external/perl/Text-Template-1.56/SIGNATURE",
        "external/perl/Text-Template-1.56/lib/Text/Template.pm",
        "external/perl/Text-Template-1.56/lib/Text/Template/Preprocess.pm",
        "external/perl/Text-Template-1.56/t/author-pod-syntax.t",
        "external/perl/Text-Template-1.56/t/author-signature.t",
        "external/perl/Text-Template-1.56/t/basic.t",
        "external/perl/Text-Template-1.56/t/broken.t",
        "external/perl/Text-Template-1.56/t/delimiters.t",
        "external/perl/Text-Template-1.56/t/error.t",
        "external/perl/Text-Template-1.56/t/exported.t",
        "external/perl/Text-Template-1.56/t/hash.t",
        "external/perl/Text-Template-1.56/t/inline-comment.t",
        "external/perl/Text-Template-1.56/t/nested-tags.t",
        "external/perl/Text-Template-1.56/t/ofh.t",
        "external/perl/Text-Template-1.56/t/out.t",
        "external/perl/Text-Template-1.56/t/prepend.t",
        "external/perl/Text-Template-1.56/t/preprocess.t",
        "external/perl/Text-Template-1.56/t/rt29928.t",
        "external/perl/Text-Template-1.56/t/safe.t",
        "external/perl/Text-Template-1.56/t/safe2.t",
        "external/perl/Text-Template-1.56/t/safe3.t",
        "external/perl/Text-Template-1.56/t/strict.t",
        "external/perl/Text-Template-1.56/t/taint.t",
        "external/perl/Text-Template-1.56/t/template-encoding.t",
        "external/perl/Text-Template-1.56/t/warnings.t",
    ] + CRYPTO_SRCS + TEMPLATED_OPENSSL_HEADERS,
)
```

This produces the following result in the console:

```
Usage: Configure [no-<feature> ...] [enable-<feature> ...] [-Dxxx] [-lxxx] [-Lxxx] [-fxxx] [-Kxxx] [no-hw-xxx|no-hw] [[no-]threads] [[no-]thread-pool] [[no-]default-thread-pool] [[no-]shared] [[no-]zlib|zlib-dynamic] [no-asm] [no-egd] [sctp] [386] [--prefix=DIR] [--openssldir=OPENSSLDIR] [--with-xxx[=vvv]] [--config=FILE] os/compiler[:flags]

pick os/compiler from:
/dev/third_party/openssl/openssl /dev/
Configuring OpenSSL version 3.2.0 for target linux-x86_64
Using os-specific seed configuration
```

This seems to fail on the following line: https://github.com/openssl/openssl/blob/master/Configure#L1340

I believe whats going on here is that it can't find any os/compiler pairs, and also when the `LIST` is ran it produces nothing. Is there anything I can do to give this script some hints as to the found os/compiler flags?

**Comment by levitte** [comment, marked as answer]
`Configure` (and `config`) also depend on the files in the `Configurations/` subdirectory.  If you include those, `LIST` will give you a better answer.

**Comment by 8W9aG** [comment]
That was it thanks! Would openssl be open to me making a PR that sends an explicit error from the `Configure` script when no Configurations are present? It seems to me it can't meaningfully progress beyond that point and it would be great to have that as an explicit error.

Simply changing the rule to incorporate the entire folder containing openssl did indeed work, now looks like so:

```
genrule(
    name = "configure",
    outs = GENERATED_OPENSSL_HEADERS,
    cmd = OPENSSL_CONFIG_CMD,
    tools = glob(["**"]),
)
```

Something I've noticed is the `include/openssl/configuration.h` file gets generated during the configuration but other files like `include/openssl/err.h` get generated during the `make` command. This is in general ok but I would rather have bazel build these objects for build efficiency purposes/hermeticity. Is there a way to get the make command to generate just the templated files? Or perhaps a way I can call the `dofile.pl` script to generate them myself after config is called?

**Reply by levitte** [reply]
`make build_generated` will do what you ask for.

**Comment by 8W9aG** [comment]
To answer my latter question, it seems that the make file adds steps to generate these header files which I can call explicitly like so:

`make include/openssl/err.h include/openssl/opensslv.h include/openssl/conf.h`

**Reply by mspncp** [reply]
As @levitte [pointed out earlier](https://github.com/openssl/openssl/discussions/22826#discussioncomment-7684119), the target `build_generated` can be used to build all of those generated files at once.

**Comment by mattcaswell** [comment]
>  Would openssl be open to me making a PR that sends an explicit error from the Configure script when no Configurations are present? 

Sure. Please feel free to open a PR.
